R/metab_simple.R

```diff
@@ -48,31 +54,39 @@ metab_simple <- function(data, ...) {
   #' 
   #' @param params a vector of length 2, where the first element is GPP and the second element is ER (both mg/L/d)
   #' @param 
-#   onestation_likelihood <- function(params, DO, DO.deficit, oxy, light, z, bp, ts, K) {
-#     
-#     GPP <- params[1]
-#     ER <- params[2]
-#     
-#     DO.modeled <- numeric(nrow(data))
-#     DO.modeled[1] <- oxy[1]
-#     for (i in 2:length(DO)) {
-#       DO.modeled[i] <- 
-#         DO.modeled[i-1] +
-#         (GPP/z)*(light[i]/sum(light)) + 
-#         ER * ts / z + 
-#         Kcor(temp[i],K)*ts*DO.deficit 
-#     }
-#   
-#     ##below is MLE calculation
-#     sqdiff<-(oxy-metab)^2 
-#     length(oxy)*(log(((sum(sqdiff)/length(oxy))^0.5)) +0.5*log(6.28))   + ((2*sum(sqdiff)/length(oxy))^-1)*sum(sqdiff)
-#   }
-  
+  onestation_negloglik <- function(params, DO.obs, DO.deficit, depth, k.O2, frac.GPP, frac.ER, frac.D) {
+    
+    # Parse params vector (passed from nlm)
+    GPP <- params[1]
+    ER <- params[2]
+    
+    # Model DO with given params
+    DO.mod <- numeric(length(DO.obs))
+    DO.mod[1] <- DO.obs[1]
+    DO.delta <- 
+      GPP * frac.GPP / depth + 
+      ER * frac.ER / depth + 
+      k.O2 * DO.deficit * frac.D # Bob - pros & cons of using DO.mod[t-1] instead of DO.obs[t] or DO.obs[t-1] (i.e., from DO.deficit) here?
```

Comment:
Yes, this is a known mistake with the model. By using t-1 we are incorporating an error. When dt= 5 min, then it is not a problem. But with dt =30, the error is bigger. There is a way to take the average, it involves some algebra. Let's talk about it, because I think it is the way to go, despite that I lost  4 h in my office this week trying to get it to work.

And we could use dnorm() for the mle,  but then need to add an explicit sd parameter.
Bob



On Apr 24, 2015, at 5:36 PM, Jordan S Read <notifications@github.com<mailto:notifications@github.com>> wrote:


In R/metab_simple.R<https://github.com/USGS-R/streamMetabolizer/pull/25#discussion_r29093951>:

> -#     length(oxy)*(log(((sum(sqdiff)/length(oxy))^0.5)) +0.5*log(6.28))   + ((2*sum(sqdiff)/length(oxy))^-1)*sum(sqdiff)
> -#   }
> -
> +  onestation_negloglik <- function(params, DO.obs, DO.deficit, depth, k.O2, frac.GPP, frac.ER, frac.D) {
> +
> +    # Parse params vector (passed from nlm)
> +    GPP <- params[1]
> +    ER <- params[2]
> +
> +    # Model DO with given params
> +    DO.mod <- numeric(length(DO.obs))
> +    DO.mod[1] <- DO.obs[1]
> +    DO.delta <-
> +      GPP * frac.GPP / depth +
> +      ER * frac.ER / depth +
> +      k.O2 * DO.deficit * frac.D # Bob - pros & cons of using DO.mod[t-1] instead of DO.obs[t] or DO.obs[t-1] (i.e., from DO.deficit) here?


Interesting. I hadn't thought about it that way. The trouble I am having w/ metab[i-1] is that it is timestep-sensitive. The potential difference between now and then in model-land is going to be greater with larger delta-time.

-
Reply to this email directly or view it on GitHub<https://github.com/USGS-R/streamMetabolizer/pull/25/files#r29093951>.